pkgs/top-level/all-packages.nix

```diff
@@ -139,16 +139,35 @@ with pkgs;
   ### Push NixOS tests inside the fixed point
 
   # See also allTestsForSystem in nixos/release.nix
-  nixosTests = import ../../nixos/tests/all-tests.nix {
-    inherit pkgs;
-    system = stdenv.hostPlatform.system;
+  nixosTests = let
+    systemArgs =
+      if stdenv.hostPlatform.isLinux && (stdenv.buildPlatform.isLinux || stdenv.buildPlatform.isDarwin)
+      then {
+        hostPkgs = pkgs.buildPackages;
+        pkgs = pkgs;
+        system = stdenv.hostPlatform.system;
+      }
+      else if stdenv.hostPlatform.isDarwin && stdenv.buildPlatform.isDarwin
+      then
+        # In the context of a normal darwin compile, repurpose nixosTests as a
+        # non-cross build. This requires a "remote" builder such as a local
+        # Linux VM, but all VM hosting happens on darwin.
+        let linuxSystem = "${stdenv.buildPlatform.parsed.cpu.name}-linux";
+        in {
+          hostPkgs = pkgs;
+          pkgs = import ../.. { inherit config overlays; system = linuxSystem; };
+          system = linuxSystem;
+        }
+      else throw "Don't know how to configure NixOS for VM host ${stdenv.buildPlatform.system} and/or VM guest ${stdenv.hostPlatform.system} built on ${stdenv.buildPlatform.system}";
```

Comment:
This logic is useful in other places, so it should be factored or moved to a common location, such as the nixos/lib/testing/pkgs.nix module.

Reply:
How about `pkgsLinux` (in `pkgs`)